### PR TITLE
MULE-9397: Add support for OnException on Sources

### PR DESCRIPTION
**Pull request by juandesi**

### CODE REVIEW

File: modules/extensions-spring-support/src/test/java/org/mule/module/extension/internal/MessageSourceTestCase.java [hunk 1]
```diff
@@ -52,6 +57,21 @@ public void onException() throws Exception
         new PollingProber(TIMEOUT_MILLIS, POLL_DELAY_MILLIS).check(new JUnitLambdaProbe(() -> heisenberg.getMoney().longValue() == -1));
     }
 
+
+    @Test
+    public void enrichExceptionOnStart() throws Exception
+    {
+        Throwable exception = catchThrowable(() -> startFlow("sourceFailedOnStart"));
+        assertThat(exception.getMessage(), StringContains.containsString(HeisenbergConnectionExceptionEnricher.ENRICHED_MESSAGE + HeisenbergSource.CORE_POOL_SIZE_ERROR_MESSAGE));
```
Comment (marianogonzalez) [inline modules/extensions-spring-support/src/test/java/org/mule/module/extension/internal/MessageSourceTestCase.java]: static imports... static imports everywhere...

File: modules/extensions-spring-support/src/test/java/org/mule/module/extension/internal/OperationExecutionTestCase.java [hunk 2]
```diff
@@ -268,7 +269,7 @@ public void operationWhichRequiresConnection() throws Exception
     public void extensionWithExceptionEnricher() throws Throwable
     {
         expectedException.expect(ConnectionException.class);
-        expectedException.expectMessage(is(HeisenbergOperations.CALL_GUS_MESSAGE));
+        expectedException.expectMessage(is(HeisenbergConnectionExceptionEnricher.ENRICHED_MESSAGE + HeisenbergOperations.CALL_GUS_MESSAGE));
```
Comment (marianogonzalez) [inline modules/extensions-spring-support/src/test/java/org/mule/module/extension/internal/OperationExecutionTestCase.java]: static imports

File: modules/extensions-support/src/main/java/org/mule/module/extension/internal/runtime/source/ExtensionMessageSource.java [hunk 3]
```diff
@@ -382,4 +385,27 @@ private void notifyExceptionAndShutDown(Throwable exception)
         LOGGER.error(String.format("Message source '%s' on flow '%s' threw exception. Shutting down it forever...", source.getName(), flowConstruct.getName()), exception);
         shutdown();
     }
+
+    private Exception enrichException(Throwable t)
+    {
+
+        Optional<ExceptionEnricherFactory> exceptionEnricherFactory = sourceModel.getExceptionEnricherFactory();
```
Comment (marianogonzalez) [inline modules/extensions-support/src/main/java/org/mule/module/extension/internal/runtime/source/ExtensionMessageSource.java]: don't do this everytime. Obtain the enricher once and reuse

File: modules/extensions-support/src/main/java/org/mule/module/extension/internal/runtime/source/ExtensionMessageSource.java [hunk 4]
```diff
@@ -382,4 +385,27 @@ private void notifyExceptionAndShutDown(Throwable exception)
...
+        if (!exceptionEnricherFactory.isPresent())
+        {
+            exceptionEnricherFactory = extensionModel.getExceptionEnricherFactory();
+        }
+
+        Exception e;
```
Comment (marianogonzalez) [inline modules/extensions-support/src/main/java/org/mule/module/extension/internal/runtime/source/ExtensionMessageSource.java]: reuse the code at org.mule.module.extension.internal.runtime.DefaultExecutionMediator#processException instead. A refactor might be needed to make that code shareable

File: modules/extensions-support/src/test/java/org/mule/module/extension/internal/runtime/source/ExtensionMessageSourceTestCase.java [hunk 5]
```diff
@@ -287,6 +300,33 @@ public void stop() throws Exception
...
     @Test
+    public void enrichExceptionWithSourceExceptionEnricher() throws Exception
+    {
+        when(enricherFactory.createEnricher()).thenReturn(new HeisenbergConnectionExceptionEnricher());
+        when(sourceModel.getExceptionEnricherFactory()).thenReturn(Optional.of(enricherFactory));
+        doThrow(new RuntimeException(ERROR_MESSAGE)).when(source).start();
+        Throwable t = catchThrowable(messageSource::start);
+
+        assertThat(ExceptionUtils.containsType(t, ConnectionException.class), is(true));
+        assertThat(t.getMessage(), StringContains.containsString(HeisenbergConnectionExceptionEnricher.ENRICHED_MESSAGE + ERROR_MESSAGE));
```
Comment (marianogonzalez) [inline modules/extensions-support/src/test/java/org/mule/module/extension/internal/runtime/source/ExtensionMessageSourceTestCase.java]: static imports

File: modules/extensions-spring-support/src/test/java/org/mule/module/extension/internal/MessageSourceTestCase.java [hunk 6]
```diff
@@ -52,6 +58,21 @@ public void onException() throws Exception
         new PollingProber(TIMEOUT_MILLIS, POLL_DELAY_MILLIS).check(new JUnitLambdaProbe(() -> heisenberg.getMoney().longValue() == -1));
     }
 
+
+    @Test
+    public void enrichExceptionOnStart() throws Exception
+    {
+        Throwable exception = catchThrowable(() -> startFlow("sourceFailedOnStart"));
```
Comment (estebanwasinger) [inline modules/extensions-spring-support/src/test/java/org/mule/module/extension/internal/MessageSourceTestCase.java]: I don't think to be necessary to catch the throwable, the only assertion is over the Exception it self. Maybe you can use the ExpectedException JUnit rule to check the type of the exception, the cause and the message.

**Pull request merged**